cuml/dbscan_demo.ipynb

```diff
@@ -4,174 +4,206 @@
    "metadata": {},
    "outputs": [],
    "source": [
+    "import os\n",
+    "\n",
     "import numpy as np\n",
+    "\n",
+    "from sklearn import datasets\n",
+    "\n",
     "import pandas as pd\n",
+    "import cudf as gd\n",
+    "\n",
+    "from cuml.datasets import make_blobs\n",
+    "\n",
+    "from sklearn.metrics import adjusted_rand_score\n",
+    "\n",
     "from sklearn.cluster import DBSCAN as skDBSCAN\n",
-    "from cuml import DBSCAN as cumlDBSCAN\n",
-    "import cudf\n",
-    "import os"
+    "from cuml.cluster import DBSCAN as cumlDBSCAN"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# Helper Functions"
+    "## Generate Data"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 30,
    "metadata": {},
    "outputs": [],
    "source": [
-    "# check if the mortgage dataset is present and then extract the data from it, else just create a random dataset for clustering \n",
-    "import gzip\n",
-    "# change the path of the mortgage dataset if you have saved it in a different directory\n",
-    "def load_data(nrows, ncols, cached = 'data/mortgage.npy.gz'):\n",
-    "    if os.path.exists(cached):\n",
-    "        print('use mortgage data')\n",
-    "        with gzip.open(cached) as f:\n",
-    "            X = np.load(f)\n",
-    "        X = X[np.random.randint(0,X.shape[0]-1,nrows),:ncols]\n",
-    "    else:\n",
-    "        # create a random dataset\n",
-    "        print('use random data')\n",
-    "        X = np.random.rand(nrows,ncols)\n",
-    "    df = pd.DataFrame({'fea%d'%i:X[:,i] for i in range(X.shape[1])})\n",
-    "    return df"
+    "n_samples = 100000\n",
+    "n_features = 128"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 31,
    "metadata": {},
    "outputs": [],
    "source": [
-    "# this function checks if the results obtained from two different methods is the same\n",
-    "from sklearn.metrics import mean_squared_error\n",
-    "def array_equal(a,b,threshold=5e-3,with_sign=True):\n",
-    "    a = to_nparray(a)\n",
-    "    b = to_nparray(b)\n",
-    "    if with_sign == False:\n",
-    "        a,b = np.abs(a),np.abs(b)\n",
-    "    res = mean_squared_error(a,b)<threshold\n",
-    "    return res\n",
+    "device_data, device_labels = make_blobs(\n",
+    "   n_samples=n_samples, n_features=n_features, centers=5, random_state=7)\n",
     "\n",
-    "# the function converts a variable from ndarray or dataframe format to numpy array\n",
-    "def to_nparray(x):\n",
-    "    if isinstance(x,np.ndarray) or isinstance(x,pd.DataFrame):\n",
-    "        return np.array(x)\n",
-    "    elif isinstance(x,np.float64):\n",
-    "        return np.array([x])\n",
-    "    elif isinstance(x,cudf.DataFrame) or isinstance(x,cudf.Series):\n",
-    "        return x.to_pandas().values\n",
-    "    return x"
+    "device_data = gd.DataFrame.from_gpu_matrix(device_data)\n",
```

Comment:
As a general point, for row-major algorithms, wouldn't a gpuarray be faster than a cudf?
It might be nice to point that out or have at least one demo that uses gpuarray as input.
(could wait until post 0.9 if it's more work though)

Reply:
Actually, the dbscan does not need to be row-major anymore now that our distance prim supports col-major inputs. I’ll make a ticket to change it. 

IMO, I prefer the use of cudf consistently in these notebooks since it demonstrates an end-to-end RAPIDS ecosystem. 

I’ll have to take a look at our implementation of ‘make_blobs’ but I don’t see a reason it wouldn’t be returning col-major. This means dbscan would need a transpose either way. Our transpose uses shared memory and it’s really fast. 

That being said, it would be nice to have  a notebook (maybe for 0.10?) that demonstrates one of our models accepting/returning different input types in addition to cudf, like numba, cupy, numpy, etc.. we could then link to that notebook from all of the model demo notebooks. 

I’ve been thinking it would also be nice to have a notebook demonstrating the loading and saving of models. 

What do you think?

Reply:
I'm ok with a separate notebook demonstrating alternative input formats.